util/cpuprofile/cpuprofile.go

```diff
@@ -0,0 +1,243 @@
+// Normally, the registered ProfileConsumer will receive the cpu profile data per second.
+// If the ProfileConsumer (channel) is full, the latest cpu profile data will not be sent to it.
+// This function is thread-safe.
+func Register(ch ProfileConsumer) {
+	globalCPUProfiler.register(ch)
+}
+
+// Unregister unregister a ProfileConsumer from the global CPU profiler.
+// The unregistered ProfileConsumer won't receive the cpu profile data any more.
+// This function is thread-safe.
+func Unregister(ch ProfileConsumer) {
+	globalCPUProfiler.unregister(ch)
+}
+
+// ConsumersCount returns the count of the global parallelCPUProfiler's consumer.It is exporting for test.
+func ConsumersCount() int {
+	return globalCPUProfiler.consumersCount()
+}
+
+// parallelCPUProfiler is a cpu profiler.
+// With parallelCPUProfiler, it is possible to have multiple profile consumer at the same time.
+// WARN: Only one running parallelCPUProfiler is allowed in the process, otherwise some profiler may profiling fail.
+type parallelCPUProfiler struct {
+	sync.Mutex
+	cs     map[ProfileConsumer]struct{}
+	notify chan struct{}
+
+	// some data cache for profiling.
+	data         *ProfileData
+	nextData     *ProfileData
+	lastDataSize int
+
+	started  atomic.Bool
+	closed   chan struct{}
+	isClosed atomic.Bool
+	wg       sync.WaitGroup
+}
+
+// newParallelCPUProfiler crate a new parallelCPUProfiler.
+func newParallelCPUProfiler() *parallelCPUProfiler {
+	return &parallelCPUProfiler{
+		cs:     make(map[ProfileConsumer]struct{}),
+		notify: make(chan struct{}),
+		closed: make(chan struct{}),
+	}
+}
+
+var (
+	errProfilerAlreadyStarted = errors.New("parallelCPUProfiler already started")
+	errProfilerAlreadyClosed  = errors.New("parallelCPUProfiler already closed")
+)
+
+func (p *parallelCPUProfiler) start() error {
+	if !p.started.CAS(false, true) {
+		return errProfilerAlreadyStarted
+	}
+	if p.isClosed.Load() {
+		return errProfilerAlreadyClosed
+	}
+
+	logutil.BgLogger().Info("parallel cpu profiler started")
+	p.wg.Add(1)
+	go util.WithRecovery(p.profilingLoop, nil)
+	return nil
+}
+
+func (p *parallelCPUProfiler) close() {
+	if p.isClosed.CAS(false, true) {
+		close(p.closed)
+	}
+	p.wg.Wait()
+}
+
+func (p *parallelCPUProfiler) register(ch ProfileConsumer) {
+	if ch == nil {
+		return
+	}
+	p.Lock()
+	p.cs[ch] = struct{}{}
+	p.Unlock()
+	// notify
+	select {
+	case p.notify <- struct{}{}:
+	default:
+	}
+}
+
+func (p *parallelCPUProfiler) unregister(ch ProfileConsumer) {
+	if ch == nil {
+		return
+	}
+	p.Lock()
+	delete(p.cs, ch)
+	p.Unlock()
+}
+
+func (p *parallelCPUProfiler) profilingLoop() {
+	checkTicker := time.NewTicker(defProfileDuration)
+	defer func() {
+		checkTicker.Stop()
+		pprof.StopCPUProfile()
+		p.wg.Done()
+	}()
+	for {
+		select {
+		case <-p.closed:
+			return
+		case <-p.notify:
+			if !p.inProfilingStatus() {
+				p.doProfiling()
+			}
+		case <-checkTicker.C:
+			p.doProfiling()
+		}
+	}
+}
+
+func (p *parallelCPUProfiler) doProfiling() {
```

Comment:
The name is misleading. A `doXxx` function in Golang usually means it does something and waits for it. But this function is not. Please either change your API name or change or implementation.